test: Implement TC acknowledgement test cases

swan-amazon opened the pull request:
Implements test cases for Terms & Conditions acknowledgement verification:
- TC-*-2.5: SetTCAcknowledgements verification
- TC-*-2.6: CommissioningComplete with no terms accepted
- TC-*-2.7: CommissioningComplete with invalid terms
- TC-*-2.8: TCAcknowledgements reset after Factory Reset
- TC-*-2.9: Reset after fabric removal
- TC-*-2.10: Required terms validation
- TC-*-2.11: Post-commission updates

#### Testing

Test cases implemented according to test plan specifications. Each test verifies specific TC acknowledgement behaviours.

swan-amazon committed:
test: Implement TC acknowledgement test cases 2.5-2.8

Implements test cases for Terms & Conditions acknowledgement verification:
- TC-*-2.5: SetTCAcknowledgements verification
- TC-*-2.6: CommissioningComplete with no terms accepted
- TC-*-2.7: CommissioningComplete with invalid terms
- TC-*-2.8: TCAcknowledgements reset after Factory Reset

Remaining test cases to be implemented in follow-up changes:
- TC-*-2.9: Reset after fabric removal
- TC-*-2.10: Required terms validation
- TC-*-2.11: Post-commission updates

Testing:
Test cases implemented according to test plan specifications.
Each test verifies specific TC acknowledgement behaviors.

cjandhyala commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,157 @@
+        tcAcknowledgements = response[ROOT_ENDPOINT_ID][Clusters.GeneralCommissioning][Clusters.GeneralCommissioning.Attributes.TCAcknowledgements]
+        tcAcknowledgementsRequired = response[ROOT_ENDPOINT_ID][Clusters.GeneralCommissioning][Clusters.GeneralCommissioning.Attributes.TCAcknowledgementsRequired]
+
+        print(f'tcAcceptedVersion= {tcAcceptedVersion}')
```
instead of print , can you use logging , so that when this test is run this info is printed in the log file.

swan-amazon replied:
I removed the print statements.

cjandhyala reviewed (commented):
Pls make sure steps in the script matches with the steps in test plan so that we have clear traceability.

swan-amazon committed:
test: Implement remaining TC acknowledgement test cases 2.9-2.11

Implements remaining test cases for Terms & Conditions acknowledgement verification:
- TC-*-2.9: TCAcknowledgements reset after fabric removal
- TC-*-2.10: Required terms validation
- TC-*-2.11: Post-commission TC updates

This completes the test coverage for TC acknowledgement verification,
following up on the previous implementation of test cases 2.5-2.8.
The new test cases verify:
- TC state after fabric removal
- Protection of required terms
- Ability to update TC version and acknowledgements post-commissioning

Testing:
Test cases implemented according to test plan specifications.
Each test verifies specific TC acknowledgement behaviors.

swan-amazon committed:
wip

swan-amazon committed:
Merge remote-tracking branch 'origin/master' into pr/swan-amazon/37015

swan-amazon committed:
feat(testing): Add PICS guard for Terms & Conditions test steps

- Wrap test steps in PICS guard checks for "CGEN.S" and "CGEN.S.F00(TC)"
- Ensure test steps only execute when Terms & Conditions feature flag is enabled
- Maintain test logic while adding conditional execution based on PICS support

swan-amazon commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,157 @@
+
+    @async_test_body
+    async def test_TC_CGEN_2_5(self):
+        commissioner: ChipDeviceCtrl.ChipDeviceController = self.default_controller
```
Updated to skip all steps when the PICS feature flag is missing.

tehampson replied:
I asked Cecille about this as I thought this check was odd, here is what she has to say when I asked if there should be an explicit skip all tests with a PICS check in the test itself:

>no - they should be adding that pics as a top-level pics (so the TH says it doesn't need to be run). And if they want to be able to run the test against any device without first checking the TC feature, they should change the decorator on the test to use @run_if_endpoint_matches(has_command(whatever that command is)) or has_feature. Either way. Using the decorator means the entire test gets marked as being skipped, and the TH will not indicate the test has been run when it hasn't. 

My 2 cents is that we should be relying on top-level pics and not this:
```
        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
            self.skip_all_remaining_steps(1)
            return
```

swan-amazon replied:
That only works for tests that begin after the device is pre-commissioned, unless I don't understand.

tehampson replied:
Marking unresolved again. I just want to double check, after the discussion in the meeting it sounded like there is something we should be doing here instead.

swan-amazon replied:
Updated the skip strategy.

swan-amazon commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,157 @@
+
+        # Don't set TCs for the next commissioning and skip CommissioningComplete so we can manually call CommissioningComplete in order to check the response error code
+        self.step(1)
+        commissioner.SetTCRequired(False)
```
This is what we intended. Chris Koos had some feedback that drove this specific text. The `commission_devices()` function isn't that great... We need to make sure the `commissioning_method` property is not present when the `default_matter_test_main` function is called, otherwise, internally, it adds the "CommissionDeviceTest" class to the test case, and then calls to `commission_devices()` will auto commission through completion, unless we set the skip `SetSkipCommissioningComplete` property on the AutoCommissioner.

swan-amazon commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,157 @@
+        commissioner.SetTCRequired(False)
+        commissioner.SetSkipCommissioningComplete(True)
+        self.matter_test_config.commissioning_method = self.matter_test_config.in_test_commissioning_method
+        await self.commission_devices()
```
That conversation appears to be related to re-using commissioning data when the certification tool is being used as the front end?

We need a better utility function to perform commissioning through specific stages... But hopefully that won't block the current test scripts from being integrated for v1.4.1.

swan-amazon committed:
Added PICS and PIXIT definitions to src/app/tests/suites/certification/PICS.yaml

swan-amazon committed:
Merge remote-tracking branch 'origin/master' into pr/swan-amazon/37015

swan-amazon committed:
refactor: Extract commission_devices functionality into reusable methods

Move commissioning logic from CommissionDeviceTest class into standalone
functions to improve code reusability. Introduce CommissioningInfo dataclass
to encapsulate commissioning parameters and add commission_device/
commission_devices helper functions.

The changes:
- Move SetupPayloadInfo dataclass to module level
- Add new CommissioningInfo dataclass for commissioning parameters
- Extract commission_device and commission_devices as standalone async functions
- Add commission_devices method to MatterBaseTest class
- Maintain backward compatibility with existing usage in CommissionDeviceTest

koosc commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,198 @@
+            TestStep(8, "TH sends CommissioningComplete to DUT."),
+            TestStep(9, "TH reads from the DUT the attribute TCAcceptedVersion."),
+            TestStep(10, "TH reads from the DUT the attribute TCAcknowledgements."),
+            TestStep(11, "TH reads from the DUT the attribute."),
```
Missing `TCMinRequiredVersion`. This was a mistake in the test plan that I'll open a PR for.

swan-amazon replied:
Added.

koosc replied:
PR for fixing this: https://github.com/CHIP-Specifications/chip-test-plans/pull/4899

koosc commented on src/python_testing/TC_CGEN_2_10.py:
```diff
@@ -0,0 +1,158 @@
+        self.matter_test_config.tc_user_response_to_simulate = None
+        await self.commission_devices()
+
+        # await self.commission_devices()
```
Should this be removed?

swan-amazon replied:
Thx. Removed.

swan-amazon committed:
Merge branch 'master' into testing/tc-test-plan-scripts-1

swan-amazon committed:
Merge remote-tracking branch 'origin/master' into pr/swan-amazon/37015

swan-amazon committed:
Addressed PR comments

swan-amazon committed:
Merge remote-tracking branch 'origin/master' into pr/swan-amazon/37015

swan-amazon committed:
Merge remote-tracking branch 'origin/master' into pr/swan-amazon/37015

tehampson commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,197 @@
+        if isinstance(tc_update_deadline, ValueDecodeFailure):
+            asserts.assert_is_none(tc_update_deadline.TLVValue)
+        else:
+            asserts.assert_less(tc_update_deadline, 2**32, "TCUpdateDeadline exceeds uint32 range")
```
Why not use asserts from `src/python_testing/matter_testing_infrastructure/chip/testing/matter_asserts.py` There is `assert_valid_uint32`.

For example you didn't check that value was <=0

swan-amazon replied:
Sure. Updated.

tehampson commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,197 @@
+        tc_update_deadline = response[ROOT_ENDPOINT_ID][Clusters.GeneralCommissioning][Clusters.GeneralCommissioning.Attributes.TCUpdateDeadline]
+
+        # Validate the value is of type Optional[uint32], e.g. either None or within the 32-bit range.
+        if isinstance(tc_update_deadline, ValueDecodeFailure):
```
I am confused I though if you read an optional that isn't there you get a value of None. Why would it be of type `ValueDecodeFailure`. I could be remembering this wrong

tehampson replied:
What was the `ValueDecodeFailure`? You should see something like `TLVValue=None, Reason=InteractionModelError(<Status.UnsupportedCluster: 195>))`

tehampson replied:
Using the all-cluster app if I do the following:
```
await dev_ctrl.WriteAttribute(0x12344321, [(1, Clusters.UnitTesting.Attributes.NullableInt32u)])
foobar = await devCtrl.ReadAttribute(nodeid=0x12344321, attributes=[(1, Clusters.UnitTesting.Attributes.NullableInt32u)])
print(foobar)
```

I get:
```
{1: {<class 'chip.clusters.Objects.UnitTesting'>: {<class 'chip.clusters.Attribute.DataVersion'>: 1494564830, <class 'chip.clusters.Objects.UnitTesting.Attributes.NullableInt32u'>: Null}}}
```

I think if you are expecting null here you should be check if it equal `NullValue` from `from chip.clusters.Types import NullValue`

Please ping me on slack if you have any questions

tehampson replied:
There actually is a larger problem:
https://github.com/project-chip/connectedhomeip/blob/master/src/app/zap-templates/zcl/data-model/chip/general-commissioning-cluster.xml#L108

The [spec mentions](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/service_device_management/GeneralCommissioningCluster.adoc?plain=1#L222) says that this attribute is nullable (`X`), but in the [xml it is marked as](https://github.com/project-chip/connectedhomeip/blob/master/src/app/zap-templates/zcl/data-model/chip/general-commissioning-cluster.xml#L108) `optional="true"` when it should be `isNullable="true"`

tehampson replied:
If spec is right and this should be nullable I created this PR in hopes to help speed up this fix in time for Friday
https://github.com/project-chip/connectedhomeip/pull/37438

cecille replied:
Agreed - value decode failure means the app isn't returning something correctly. If it's optional and not implemented, you should be getting an unsupported attribute error

swan-amazon replied:
I was able to update the check with a "isinstace(chip.clusters.Types.Nullable)"

tehampson commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,197 @@
+            attributes=[(ROOT_ENDPOINT_ID, Clusters.GeneralCommissioning.Attributes.FeatureMap)],
+        )
+        feature_map = response[ROOT_ENDPOINT_ID][Clusters.GeneralCommissioning][Clusters.GeneralCommissioning.Attributes.FeatureMap]
+        asserts.assert_equal(feature_map & 0x1, 0x1, "TC feature flag is not set.")
```
Why not use `Clusters.GeneralCommissioning.Bitmaps.Feature.kTermsAndConditions`. This is more readable

swan-amazon replied:
Thanks! that is better

tehampson commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,197 @@
+        self.step(11)
+        response = await commissioner.ReadAttribute(nodeid=self.dut_node_id, attributes=[(ROOT_ENDPOINT_ID, Clusters.GeneralCommissioning.Attributes.TCMinRequiredVersion)])
+        min_required_version = response[ROOT_ENDPOINT_ID][Clusters.GeneralCommissioning][Clusters.GeneralCommissioning.Attributes.TCMinRequiredVersion]
+        asserts.assert_is_instance(min_required_version, int, "TCMinRequiredVersion is not a uint16 type.")
```
Please use `assert_valid_uint16` from matter_assert

swan-amazon replied:
OK, let me know if I missed anywhere.

tehampson commented on src/python_testing/TC_CGEN_2_8.py:
```diff
@@ -0,0 +1,164 @@
+            self.skip_all_remaining_steps(7)
+            return
+
+        self.wait_for_user_input(prompt_msg="Set the DUT into commissioning mode")
```
Might be helpful to anyone testing for this prompt to mention that the DUT should be factory resetting. The more information you provided to a test that is testing this out the less friction you will encounter later at certification

swan-amazon replied:
I'm updating the script to prompt the user to manually trigger factory reset and manually set the device in commissioning mode, then asking to "continue", e.g. only press OK afterward.

tehampson commented on src/python_testing/TC_CGEN_2_9.py:
```diff
@@ -0,0 +1,194 @@
+        await self.remove_commissioner_fabric()
+
+        # Step 6: Put device in commissioning mode (requiring user input, so skip in CI)
+        self.step(6)
+        if not self.check_pics('PICS_USER_PROMPT'):
+            self.skip_all_remaining_steps(7)
+            return
```
After you remove all fabrics doesn't the example app automatically go into a commisionable state. If so you can check to see if this is CI and skip the `wait_for_user_input` if it is CI

swan-amazon replied:
No. Unfortunately it does not. I validated this behavior against the all-clusters-app (and also the terms-and-conditions-app).

Someone (maybe Tennessee) mentioned something similar on a previous PR. Did the app used to do that? Was there a regression?

```
[1738714949.583433][13:13] CHIP:EM: >>> [E:7980r S:64720 M:46214784] (S) Msg RX from 1:000000000001B669 [5C33] to 0000000000000001 --- Type 0001:08 (IM:InvokeCommandRequest) (B:62)
[1738714949.583456][13:13] CHIP:EM: Handling via exchange: 7980r, Delegate: 0x55702821b0a8
[1738714949.583492][13:13] CHIP:DMG: InvokeRequestMessage =
[1738714949.583508][13:13] CHIP:DMG: {
[1738714949.583515][13:13] CHIP:DMG:    suppressResponse = false, 
[1738714949.583521][13:13] CHIP:DMG:    timedRequest = false, 
[1738714949.583525][13:13] CHIP:DMG:    InvokeRequests =
[1738714949.583533][13:13] CHIP:DMG:    [
[1738714949.583541][13:13] CHIP:DMG:            CommandDataIB =
[1738714949.583549][13:13] CHIP:DMG:            {
[1738714949.583562][13:13] CHIP:DMG:                    CommandPathIB =
[1738714949.583573][13:13] CHIP:DMG:                    {
[1738714949.583586][13:13] CHIP:DMG:                            EndpointId = 0x0,
[1738714949.583594][13:13] CHIP:DMG:                            ClusterId = 0x3e,
[1738714949.583606][13:13] CHIP:DMG:                            CommandId = 0xa,
[1738714949.583616][13:13] CHIP:DMG:                    },
[1738714949.583634][13:13] CHIP:DMG: 
[1738714949.583642][13:13] CHIP:DMG:                    CommandFields = 
[1738714949.583649][13:13] CHIP:DMG:                    {
[1738714949.583657][13:13] CHIP:DMG:                            0x0 = 1 (unsigned), 
[1738714949.583665][13:13] CHIP:DMG:                    },
[1738714949.583671][13:13] CHIP:DMG:            },
[1738714949.583687][13:13] CHIP:DMG: 
[1738714949.583694][13:13] CHIP:DMG:    ],
[1738714949.583706][13:13] CHIP:DMG: 
[1738714949.583718][13:13] CHIP:DMG:    InteractionModelRevision = 11
[1738714949.583727][13:13] CHIP:DMG: },
[1738714949.583780][13:13] CHIP:DMG: AccessControl: checking f=1 a=c s=0x000000000001B669 t= c=0x0000_003E e=0 p=a r=i
[1738714949.583795][13:13] CHIP:DMG: AccessControl: allowed
[1738714949.583805][13:13] CHIP:DMG: Received command for Endpoint=0 Cluster=0x0000_003E Command=0x0000_000A
[1738714949.583821][13:13] CHIP:ZCL: OpCreds: Received a RemoveFabric Command for FabricIndex 0x1
[1738714949.583875][13:13] CHIP:EM: <<< [E:7980r S:64720 M:71691742 (Ack:46214784)] (S) Msg TX from 0000000000000001 to 1:000000000001B669 [5C33] [UDP:[fe80::816:6bff:fe59:6fe9%eth0]:55506] --- Type 0000:10 (SecureChannel:StandaloneAck) (B:34)
[1738714949.583916][13:13] CHIP:EM: Flushed pending ack for MessageCounter:46214784 on exchange 7980r
[1738714949.583945][13:13] CHIP:EVL: LogEvent event number: 0x0000000000000005 priority: 1, endpoint id:  0x0 cluster id: 0x0000_0028 event id: 0x2 Epoch timestamp: 0x00000194D37CA7CF
[1738714949.583964][13:13] CHIP:DMG: All ReadHandler-s are clean, clear GlobalDirtySet
[1738714949.587398][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.590910][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.594523][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.598157][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.601910][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.605235][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.605266][13:13] CHIP:FP: Fabric (0x1) deleted.
[1738714949.605278][13:13] CHIP:ZCL: UserLabel: Last Fabric index 0x1 was removed
[1738714949.605318][13:13] CHIP:ZCL: OpCreds: Fabric index 0x1 was removed
[1738714949.605330][13:13] CHIP:DIS: Updating services using commissioning mode 0
[1738714949.606477][13:13] CHIP:DIS: CHIP minimal mDNS started advertising.
[1738714949.609469][13:13] CHIP:DMG: Endpoint 0, Cluster 0x0000_003E update version to 7785dbb6
[1738714949.609484][13:13] CHIP:DMG: Endpoint 0, Cluster 0x0000_003E update version to 7785dbb7
[1738714949.612699][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.616044][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.619408][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.623411][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.627435][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.631344][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.635249][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.638584][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.641701][13:13] CHIP:DL: Wrote settings to /tmp/chip_kvs
[1738714949.641747][13:13] CHIP:ZCL: OpCreds: RemoveFabric successful
[1738714949.641769][13:13] CHIP:DMG: Command handler moving to [NewRespons]
[1738714949.641785][13:13] CHIP:DMG: Command handler moving to [ Preparing]
[1738714949.641804][13:13] CHIP:DMG: Command handler moving to [AddingComm]
[1738714949.641821][13:13] CHIP:DMG: Command handler moving to [AddedComma]
[1738714949.641838][13:13] CHIP:IN: Expiring all sessions for fabric 0x1!!
[1738714949.641856][13:13] CHIP:IN: SecureSession[0x55705f42cf30]: MarkForEviction Type:2 LSID:64718
[1738714949.641865][13:13] CHIP:SC: SecureSession[0x55705f42cf30, LSID:64718]: State change 'kActive' --> 'kPendingEviction'
[1738714949.641874][13:13] CHIP:IN: SecureSession[0x55705f42cf30]: Released - Type:2 LSID:64718
[1738714949.641883][13:13] CHIP:IN: SecureSession[0x55705f438bb0]: MarkForEviction Type:2 LSID:64719
[1738714949.641900][13:13] CHIP:SC: SecureSession[0x55705f438bb0, LSID:64719]: State change 'kActive' --> 'kPendingEviction'
[1738714949.641917][13:13] CHIP:IN: SecureSession[0x55705f438bb0]: Released - Type:2 LSID:64719
[1738714949.641927][13:13] CHIP:IN: SecureSession[0x55705f429bd0]: MarkForEviction Type:2 LSID:64720
[1738714949.641932][13:13] CHIP:SC: SecureSession[0x55705f429bd0, LSID:64720]: State change 'kActive' --> 'kPendingEviction'
[1738714949.641957][13:13] CHIP:DMG: Decreasing reference count for CommandHandlerImpl, remaining 1
[1738714949.641973][13:13] CHIP:DMG: Decreasing reference count for CommandHandlerImpl, remaining 0
[1738714949.641986][13:13] CHIP:DMG: Command handler moving to [AwaitingDe]
[1738714949.642045][13:13] CHIP:EM: <<< [E:7980r S:64720 M:71691743 (Ack:46214784)] (S) Msg TX from 0000000000000001 to 1:000000000001B669 [0000] [UDP:[fe80::816:6bff:fe59:6fe9%eth0]:55506] --- Type 0001:09 (IM:InvokeCommandResponse) (B:70)
[1738714949.642112][13:13] CHIP:EM: ??1 [E:7980r S:64720 M:71691743] (S) Msg Retransmission to 1:000000000001B669 in 377ms [State:Active II:500 AI:300 AT:4000]
[1738714949.642132][13:13] CHIP:DMG: Command response sender moving to [AllInvokeR]
[1738714949.642192][13:13] CHIP:EM: >>> [E:7979r S:0 M:169072461 (Ack:161580762)] (U) Msg RX from 0:87D5EA5A3BFF00DE [0000] to 0000000000000000 --- Type 0000:10 (SecureChannel:StandaloneAck) (B:26)
[1738714949.642203][13:13] CHIP:EM: Found matching exchange: 7979r, Delegate: (nil)
[1738714949.642217][13:13] CHIP:EM: Rxd Ack; Removing MessageCounter:161580762 from Retrans Table on exchange 7979r
[1738714949.642911][13:13] CHIP:EM: >>> [E:7980r S:64720 M:46214785 (Ack:71691743)] (S) Msg RX from 1:000000000001B669 [0000] to 0000000000000001 --- Type 0000:10 (SecureChannel:StandaloneAck) (B:34)
[1738714949.642924][13:13] CHIP:EM: Found matching exchange: 7980r, Delegate: (nil)
[1738714949.642936][13:13] CHIP:EM: Rxd Ack; Removing MessageCounter:71691743 from Retrans Table on exchange 7980r
[1738714949.642948][13:13] CHIP:IN: SecureSession[0x55705f429bd0]: Released - Type:2 LSID:64720
```

tehampson replied:
I think there was a regression. For some reason I remember it doing that before. But also my memory could be wrong

swan-amazon committed:
Improve test assertions and manual testing steps

- Use Matter-specific type assertions in TC_CGEN_2_5
- Update TC_CGEN_2_8 manual testing flow and prompts

Rajashreekalmane commented on src/python_testing/TC_CGEN_2_5.py:
The script currently has missing validation checks for certain fields:

Step 9: The script should verify that TCAcceptedVersion fits within the uint16 range (0-65535), but this check is missing.
Step 10: The script does not validate whether TCAcknowledgements is a valid map16 type.
Step 11: The script only checks if TCMinRequiredVersion is an integer, but it should also confirm that it fits within uint16.

swan-amazon replied:
I added the missing validation that you mentioned for Step 9 and Step 10, but I think Step 11 was already doing what you are asking. Hopefully it was just a misread. Let me know though.

```
matter_asserts.assert_valid_uint16(min_required_version, "TCMinRequiredVersion is not a uint16 type.")
```

swan-amazon committed:
restyle

cecille commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,198 @@
+        tc_user_response_to_simulate = self.matter_test_config.global_test_params['PIXIT.CGEN.RequiredTCAcknowledgements']
+
+        self.step(0)
+        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
```
```suggestion
        if not self.check_pics("CGEN.S.F00"):
```

cecille commented on src/python_testing/TC_CGEN_2_5.py:
```diff
@@ -0,0 +1,198 @@
+        response = await commissioner.ReadAttribute(nodeid=self.dut_node_id, attributes=[(ROOT_ENDPOINT_ID, Clusters.GeneralCommissioning.Attributes.TCAcknowledgementsRequired)])
+        tc_acknowledgements_required = response[ROOT_ENDPOINT_ID][Clusters.GeneralCommissioning][Clusters.GeneralCommissioning.Attributes.TCAcknowledgementsRequired]
+        asserts.assert_equal(tc_acknowledgements_required, False, "TCAcknowledgementsRequired should be False.")
+
```
as above.

cecille commented on src/python_testing/TC_CGEN_2_6.py:
```diff
@@ -0,0 +1,93 @@
+
+    def pics_TC_CGEN_2_6(self) -> list[str]:
+        """ This function returns a list of PICS for this test case that must be True for the test to be run"""
+        return ["CGEN.S", "CGEN.S.F00(TC)"]
```
```suggestion
        return ["CGEN.S", "CGEN.S.F00"]
```

cecille commented on src/python_testing/TC_CGEN_2_6.py:
```diff
@@ -0,0 +1,93 @@
+        commissioner: ChipDeviceCtrl.ChipDeviceController = self.default_controller
+
+        self.step(0)
+        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
```
```suggestion
        if not self.check_pics("CGEN.S.F00"):
```

cecille commented on src/python_testing/TC_CGEN_2_6.py:
```diff
@@ -0,0 +1,93 @@
+
+        self.step(0)
+        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
+            self.skip_all_remaining_steps(1)
```
```suggestion
            asserts.skip("Skipping test as TC feature is not set")
```

cecille commented on src/python_testing/TC_CGEN_2_7.py:
```diff
@@ -0,0 +1,173 @@
+
+    def pics_TC_CGEN_2_7(self) -> list[str]:
+        """ This function returns a list of PICS for this test case that must be True for the test to be run"""
+        return ["CGEN.S", "CGEN.S.F00(TC)"]
```
```suggestion
        return ["CGEN.S", "CGEN.S.F00"]
```

cecille commented on src/python_testing/TC_CGEN_2_7.py:
```diff
@@ -0,0 +1,173 @@
+        tc_user_response_to_simulate = self.matter_test_config.global_test_params['PIXIT.CGEN.RequiredTCAcknowledgements']
+
+        self.step(0)
+        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
```
```suggestion
        if not self.check_pics("CGEN.S.F00"):
```

cecille commented on src/python_testing/TC_CGEN_2_7.py:
```diff
@@ -0,0 +1,173 @@
+
+        self.step(0)
+        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
+            self.skip_all_remaining_steps(1)
```
```suggestion
            asserts.skip("Skipping test as TC feature is not set")
```

cecille commented on src/python_testing/TC_CGEN_2_8.py:
```diff
@@ -0,0 +1,165 @@
+
+    def pics_TC_CGEN_2_8(self) -> list[str]:
+        """ This function returns a list of PICS for this test case that must be True for the test to be run"""
+        return ["CGEN.S", "CGEN.S.F00(TC)"]
```
```suggestion
        return ["CGEN.S", "CGEN.S.F00"]
```

cecille commented on src/python_testing/TC_CGEN_2_8.py:
```diff
@@ -0,0 +1,165 @@
+        tc_user_response_to_simulate = self.matter_test_config.global_test_params['PIXIT.CGEN.RequiredTCAcknowledgements']
+
+        self.step(0)
+        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
```
```suggestion
        if not self.check_pics("CGEN.S.F00"):
```

cecille commented on src/python_testing/TC_CGEN_2_8.py:
```diff
@@ -0,0 +1,165 @@
+
+        self.step(0)
+        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
+            self.skip_all_remaining_steps(1)
```
```suggestion
            asserts.skip("Skipping test as TC feature is not set")
```

cecille commented on src/python_testing/TC_CGEN_2_9.py:
```diff
@@ -0,0 +1,194 @@
+
+    def pics_TC_CGEN_2_9(self) -> list[str]:
+        """ This function returns a list of PICS for this test case that must be True for the test to be run"""
+        return ["CGEN.S", "CGEN.S.F00(TC)"]
```
```suggestion
        return ["CGEN.S", "CGEN.S.F00"]
```

cecille commented on src/python_testing/TC_CGEN_2_9.py:
```diff
@@ -0,0 +1,194 @@
+        tc_user_response_to_simulate = self.matter_test_config.global_test_params['PIXIT.CGEN.RequiredTCAcknowledgements']
+
+        self.step(0)
+        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
```
```suggestion
        if not self.check_pics("CGEN.S.F00"):
```

cecille commented on src/python_testing/TC_CGEN_2_9.py:
```diff
@@ -0,0 +1,194 @@
+
+        self.step(0)
+        if not self.pics_guard(self.check_pics("CGEN.S.F00(TC)")):
+            self.skip_all_remaining_steps(1)
```
```suggestion
            asserts.skip("Skipping test as TC feature is not set")
```

swan-amazon committed:
refactor: simplify CGEN feature identifier from CGEN.S.F00(TC) to CGEN.S.F00

- Update PICS.yaml and ci-pics-values to use simplified feature ID
- Update Python test files to use new feature identifier
- Improve error messaging when commissioning feature check fails
- Affected test files: TC_CGEN_2_5 through TC_CGEN_2_11

swan-amazon committed:
Update TC_CGEN_2_5 to use Nullable type for TCUpdateDeadline validation

swan-amazon committed:
Merge remote-tracking branch 'origin/master' into pr/swan-amazon/37015

swan-amazon commented on src/python_testing/TC_CGEN_2_11.py:
```diff
@@ -0,0 +1,179 @@
+        self.step(7)
+        response = await commissioner.ReadAttribute(nodeid=self.dut_node_id, attributes=[(ROOT_ENDPOINT_ID, Clusters.GeneralCommissioning.Attributes.TCAcknowledgements)])
+        current_acknowledgements = response[ROOT_ENDPOINT_ID][Clusters.GeneralCommissioning][Clusters.GeneralCommissioning.Attributes.TCAcknowledgements]
+        asserts.assert_equal(current_acknowledgements, 65535, "TCAcknowledgements not updated to maximum value")
```
Yeah that is correct. The commissioner is able to lookup potentially a new T&C set than the device is aware of, in which case the device is expected to accept extra bits. After the device is updated with the latest firmware then it should know how to check the extra bits.

swan-amazon committed:
Add type validation for TCAcceptedVersion and TCAcknowledgements

cecille commented on src/python_testing/TC_CGEN_2_10.py:
```diff
@@ -0,0 +1,154 @@
+            user_acknowledgements,
+            "TCAcknowledgements changed unexpectedly after second attempt",
+        )
+
```
Let's handle as a follow up - https://github.com/project-chip/matter-test-scripts/issues/501. I will get our team to handle this as a part of the tech debt cleanup. You just need to make sure the folks at SVE and at cert know they will need to factory reset between all of these tests or they will fail out in mysterious ways.

cecille commented on src/python_testing/TC_CGEN_2_11.py:
```diff
@@ -0,0 +1,179 @@
+        response = await commissioner.ReadAttribute(nodeid=self.dut_node_id, attributes=[(ROOT_ENDPOINT_ID, Clusters.GeneralCommissioning.Attributes.TCAcknowledgements)])
+        current_acknowledgements = response[ROOT_ENDPOINT_ID][Clusters.GeneralCommissioning][Clusters.GeneralCommissioning.Attributes.TCAcknowledgements]
+        asserts.assert_equal(current_acknowledgements, 65535, "TCAcknowledgements not updated to maximum value")
+
```
as above, handle as followup.
